jestSupport/preprocessor.js

```diff
@@ -26,6 +26,6 @@ module.exports = {
   getCacheKey: createCacheKeyFunction([
     __filename,
     path.join(__dirname, '../packager/transformer.js'),
-    path.join(__dirname, '../node_modules/babel-core/package.json'),
+    path.join(require.resolve('babel-core')),
```

Comment:
I'm not sure if it's a good solution (and very close to your first option) but
```js
path.join(require.resolve('babel-core'), '../package.json'),
```
works fine on `npm3`.